tmva/tmva/inc/TMVA/BatchLoader.hxx

```diff
@@ -0,0 +1,184 @@
+#ifndef TMVA_BATCHLOADER
+#define TMVA_BATCHLOADER
+
+#include <iostream>
+#include <vector>
+
+// Imports for threading
+#include <queue>
+#include <mutex>
+#include <condition_variable>
+
+#include "TMVA/RTensor.hxx"
+#include "TMVA/Tools.h"
+#include "TRandom3.h"
+
+
+namespace TMVA {
+namespace Experimental {
+
+class BatchLoader
+{
+private:
+    const size_t batch_size, num_columns;
+
+    bool accept_tasks = false;
+    TMVA::RandomGenerator<TRandom3> rng;
+
+    // filled batch elements
+    std::mutex batch_lock;
+    std::condition_variable batch_condition;
+    std::queue<TMVA::Experimental::RTensor<float>*> training_batch_queue;
+
+    std::vector<TMVA::Experimental::RTensor<float>*> validation_batches;
+    size_t valid_idx = 0;
+
+public:
+    /////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+    // Constructors
+    /////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+    BatchLoader(const size_t _batch_size, const size_t _num_columns)
+        : batch_size(_batch_size), num_columns(_num_columns)
+    {
+        rng = TMVA::RandomGenerator<TRandom3>(0);
+    }
+    
+    
+    ~BatchLoader() {}
+
+
+public:
+    /////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+    // Batch functions
+    /////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+    // return a batch of data
+    TMVA::Experimental::RTensor<float>* GetTrainBatch()
```

Comment:
Consider moving some (if not all) the member functions below to a dedicated translation unit, i.e. a `.cxx` file.